File: Java/HeapSort.java
public class HeapSort {
    // Array of integers
    static int[] arr = { 76, 22, 21, 96, 70, 20 };

    // Heap Sort
    public static void main(String[] args) {
        // Heap Sort
        heapSort(arr);
        int n=arr.length;
        for (int i = 0; i < n; ++i)
            System.out.print(arr[i] + " ");
        System.out.println();
    
    }

    static void heapSort(int[] arr) {

    int n = arr.length;

    // Build max heap
      for (int i = n / 2 - 1; i >= 0; i--) {
        heapify(arr, n, i);
    }

    // Heap sort
      for (int i = n - 1; i >= 0; i--) {
        int temp = arr[0];
        arr[0] = arr[i];
        arr[i] = temp;

        // Heapify root element
        heapify(arr, i, 0);
    }
}

    static void heapify(int arr[], int n, int i) {
        // Find largest among root, left child and right child
        int largest = i;
        int l = 2 * i + 1;
        int r = 2 * i + 2;

        if (l < n && arr[l] > arr[largest])
            largest = l;

        if (r < n && arr[r] > arr[largest])
            largest = r;

        // Swap and continue heapifying if root is not largest
        if (largest != i) {
            int swap = arr[i];
            arr[i] = arr[largest];
            arr[largest] = swap;

            heapify(arr, n, largest);
        }
    }


}

// Output:
// [20, 21, 22, 70, 76, 96]

File: Javascript/heapsort.js
// Heap Sort in JavaScript

const maxHeapify = (arr, n, i) => {
    let largest = i;
    let l = 2 * i + 1;
    let r = 2 * i + 2; 
    
     if (l < n && arr[l] > arr[largest]) {
           largest = l; 
     }

     if (r < n && arr[r] > arr[largest]) {
          largest = r; 
     }
    
     
     if (largest != i) { 
          let temp = arr[i]; 
          arr[i] = arr[largest]; 
          arr[largest] = temp; 
    
        maxHeapify(arr, n, largest); 
      } 
  }
  

function heapSort(arr) {
     var n = arr.length;
      for (let i = parseInt(n / 2 - 1); i >= 0; i--) {
        maxHeapify(arr, n, i); 
    }
 
    for (let i = n - 1; i >= 0; i--) { 
       let temp = arr[0]; 
       arr[0] = arr[i]; 
       arr[i] = temp;  
       maxHeapify(arr, i, 0); 
    } 

    return arr;
}

var arr = [76, 22, 21, 96, 70, 20];

console.log(heapSort(arr));

// Output:

// [ 20, 21, 22, 70, 76, 96 ]